### PR TITLE
add local check to dtsync

### PR DESCRIPTION
**Pull request by willscott**
graphsync syncer should check if requested dags are already fully available locally before re-transferring them.

### CODE REVIEW

**Review by ischasny — APPROVED** [review]
LGTM

**Comment by masih** [comment]
Fixes https://github.com/filecoin-project/storetheindex/issues/691

**Comment by willscott** [comment]
note that it mitigates the storetheindex issue by only downloading at the 5000-advertisement interval that synchronization happens at. that should probably be mostly enough mitigation, but it is still a different semantic from the http one.

File: dtsync/syncer.go [hunk 1]
```diff
@@ -84,3 +98,45 @@ func (s *Syncer) Sync(ctx context.Context, nextCid cid.Cid, sel ipld.Node) error
...
+	// trusted because it'll be hashed/verified on the way into the link system when fetched.
+	getMissingLs.TrustedStorage = true
+	getMissingLs.StorageReadOpener = func(lc ipld.LinkContext, l ipld.Link) (io.Reader, error) {
+		r, err := s.ls.StorageReadOpener(lc, l)
+		if err == nil {
+			// Found block read opener, so return it.
+			return r, nil
+		}
+
+		return nil, fmt.Errorf("not available locally")
```
Comment (gammazero) [inline dtsync/syncer.go:114]: nit: [Indent error flow](https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow)
```suggestion
		if err != nil {
			return nil, fmt.Errorf("not available locally: %w", err)
		}
		// Found block read opener, so return it.
		return r, nil		
```

**Review by gammazero — APPROVED** [review]

**Pull request merged**